src/operators_lazyproduct.jl

```diff
@@ -62,28 +73,21 @@ permutesystems(op::LazyProduct, perm::Vector{Int}) = LazyProduct(([permutesystem
 identityoperator(::Type{LazyProduct}, ::Type{S}, b1::Basis, b2::Basis) where S<:Number = LazyProduct(identityoperator(S, b1, b2))
 
 
-# Fast in-place multiplication
 function mul!(result::Ket{B1},a::LazyProduct{B1,B2},b::Ket{B2},alpha,beta) where {B1,B2}
-    tmp1 = Ket(a.operators[end].basis_l)
-    mul!(tmp1,a.operators[end],b,a.factor,0)
+    mul!(a.ket_l[end],a.operators[end],b,a.factor,0)
     for i=length(a.operators)-1:-1:2
-        tmp2 = Ket(a.operators[i].basis_l)
-        mul!(tmp2,a.operators[i],tmp1)
-        tmp1 = tmp2
+        mul!(a.ket_l[i],a.operators[i],a.ket_l[i+1])
     end
-    mul!(result,a.operators[1],tmp1,alpha,beta)
+    mul!(result,a.operators[1],a.ket_l[2],alpha,beta)
     return result
 end
 
 function mul!(result::Bra{B2},a::Bra{B1},b::LazyProduct{B1,B2},alpha,beta) where {B1,B2}
-    tmp1 = Bra(b.operators[1].basis_r)
-    mul!(tmp1,a,b.operators[1],b.factor,0)
+    mul!(b.bra_r[1],a,b.operators[1],b.factor,0)
     for i=2:length(b.operators)-1
-        tmp2 = Bra(b.operators[i].basis_r)
-        mul!(tmp2,tmp1,b.operators[i])
-        tmp1 = tmp2
+        mul!(b.bra_r[i],b.bra_r[i-1],b.operators[i])
     end
-    mul!(result,tmp1,b.operators[end],alpha,beta)
+    mul!(result,b.bra_r[end-1],b.operators[end],alpha,beta)
```

Comment:
Similar to above, you're not using `bra_r[end]`, so you can actually remove that from the cache.

Reply:
Also, tests fail here if `length(bra_r) == 1` as you're trying to index at `0` then.